File: src/main/java/com/procurement/notice/service/EnquiryServiceImpl.java
package com.procurement.notice.service;

import com.fasterxml.jackson.databind.JsonNode;
import com.fasterxml.jackson.databind.node.ObjectNode;
import com.procurement.notice.dao.ReleaseDao;
import com.procurement.notice.exception.ErrorException;
import com.procurement.notice.exception.ErrorType;
import com.procurement.notice.model.bpe.ResponseDto;
import com.procurement.notice.model.entity.ReleaseEntity;
import com.procurement.notice.model.tender.dto.UnsuspendTenderDto;
import com.procurement.notice.model.tender.enquiry.RecordEnquiry;
import com.procurement.notice.model.tender.record.Record;
import com.procurement.notice.utils.DateUtil;
import com.procurement.notice.utils.JsonUtil;
import java.time.LocalDateTime;
import java.util.ArrayList;
import java.util.List;
import java.util.Objects;
import java.util.Optional;
import org.springframework.stereotype.Service;

@Service
public class EnquiryServiceImpl implements EnquiryService {

    private static final String SEPARATOR = "-";
    private static final String ENQUIRY_JSON = "enquiry";
    private final ReleaseService releaseService;
    private final ReleaseDao releaseDao;
    private final JsonUtil jsonUtil;
    private final DateUtil dateUtil;

    public EnquiryServiceImpl(final ReleaseService releaseService,
                              final ReleaseDao releaseDao,
                              final JsonUtil jsonUtil,
                              final DateUtil dateUtil) {
        this.releaseService = releaseService;
        this.releaseDao = releaseDao;
        this.jsonUtil = jsonUtil;
        this.dateUtil = dateUtil;
    }

    @Override
    public ResponseDto createEnquiry(final String cpid,
                                     final String stage,
                                     final LocalDateTime releaseDate,
                                     final JsonNode data) {
        final ReleaseEntity entity = Optional.ofNullable(releaseDao.getRecordByCpIdAndStage(cpid, stage))
                .orElseThrow(() -> new ErrorException(ErrorType.DATA_NOT_FOUND));
        final RecordEnquiry enquiry = jsonUtil.toObject(RecordEnquiry.class, jsonUtil.toJson(data.get(ENQUIRY_JSON)));
        final Record release = jsonUtil.toObject(Record.class, entity.getJsonData());
        addEnquiryToTender(release, enquiry);
        release.setDate(releaseDate);
        release.setId(getReleaseId(release.getOcid()));
        releaseDao.saveRelease(releaseService.getReleaseEntity(cpid, stage, release));
        return getResponseDto(cpid, release.getOcid());
    }

    @Override
    public ResponseDto addAnswer(final String cpid,
                                 final String stage,
                                 final LocalDateTime releaseDate,
                                 final JsonNode data) {
        final ReleaseEntity entity = Optional.ofNullable(releaseDao.getRecordByCpIdAndStage(cpid, stage))
                .orElseThrow(() -> new ErrorException(ErrorType.DATA_NOT_FOUND));
        final RecordEnquiry enquiry = jsonUtil.toObject(RecordEnquiry.class, jsonUtil.toJson(data.get(ENQUIRY_JSON)));
        final Record release = jsonUtil.toObject(Record.class, entity.getJsonData());
        addAnswerToEnquiry(release, enquiry);
        release.setDate(releaseDate);
        release.setId(getReleaseId(release.getOcid()));
        releaseDao.saveRelease(releaseService.getReleaseEntity(cpid, stage, release));
        return getResponseDto(cpid, release.getOcid());
    }

    @Override
    public ResponseDto unsuspendTender(final String cpid,
                                       final String stage,
                                       final LocalDateTime releaseDate,
                                       final JsonNode data) {
        final ReleaseEntity entity = Optional.ofNullable(releaseDao.getRecordByCpIdAndStage(cpid, stage))
                .orElseThrow(() -> new ErrorException(ErrorType.DATA_NOT_FOUND));
        final Record release = jsonUtil.toObject(Record.class, entity.getJsonData());
        final UnsuspendTenderDto dto = jsonUtil.toObject(UnsuspendTenderDto.class, jsonUtil.toJson(data));
        final RecordEnquiry enquiry = dto.getEnquiry();
        addAnswerToEnquiry(release, enquiry);
        release.setDate(releaseDate);
        release.setId(getReleaseId(release.getOcid()));
        release.getTender().setStatusDetails(dto.getTender().getStatusDetails());
        release.getTender().setTenderPeriod(dto.getTenderPeriod());
        release.getTender().setEnquiryPeriod(dto.getEnquiryPeriod());
        releaseDao.saveRelease(releaseService.getReleaseEntity(cpid, stage, release));
        return getResponseDto(cpid, release.getOcid());
    }

    private void addEnquiryToTender(final Record release, final RecordEnquiry enquiry) {
        List<RecordEnquiry> enquiries = release.getTender().getEnquiries();
        if (Objects.isNull(enquiries)) {
            enquiries = new ArrayList<>();
        }
        final Optional<RecordEnquiry> enquiryOptional = enquiries.stream()
                .filter(e -> e.getId().equals(enquiry.getId()))
                .findFirst();
        if (!enquiryOptional.isPresent()) {
            enquiries.add(enquiry);
            release.getTender().setEnquiries(enquiries);
        }
    }

    private void addAnswerToEnquiry(final Record release, final RecordEnquiry enquiry) {
        final List<RecordEnquiry> enquiries = release.getTender().getEnquiries();
        final Optional<RecordEnquiry> enquiryOptional = enquiries.stream()
                .filter(e -> e.getId().equals(enquiry.getId()))
                .findFirst();
        if (enquiryOptional.isPresent()) {
            final RecordEnquiry updatableEnquiry = enquiryOptional.get();
            updatableEnquiry.setAnswer(enquiry.getAnswer());
            release.getTender().setEnquiries(enquiries);
        } else {
            throw new ErrorException(ErrorType.DATA_NOT_FOUND);
        }
    }

    private String getReleaseId(final String ocId) {
        return ocId + SEPARATOR + dateUtil.milliNowUTC();
    }

    private ResponseDto getResponseDto(final String cpid, final String ocid) {
        final ObjectNode jsonForResponse = jsonUtil.createObjectNode();
        jsonForResponse.put("cpid", cpid);
        jsonForResponse.put("ocid", ocid);
        return new ResponseDto<>(true, null, jsonForResponse);
    }
}

File: src/main/kotlin/com/procurement/notice/service/EnquiryServiceImpl.kt
package com.procurement.notice.service

import com.fasterxml.jackson.databind.JsonNode
import com.procurement.notice.dao.ReleaseDao
import com.procurement.notice.exception.ErrorException
import com.procurement.notice.exception.ErrorType
import com.procurement.notice.model.bpe.ResponseDto
import com.procurement.notice.model.ocds.Tag
import com.procurement.notice.model.tender.dto.UnsuspendTenderDto
import com.procurement.notice.model.tender.enquiry.RecordEnquiry
import com.procurement.notice.model.tender.ms.Ms
import com.procurement.notice.model.tender.record.Record
import com.procurement.notice.utils.createObjectNode
import com.procurement.notice.utils.milliNowUTC
import com.procurement.notice.utils.toJson
import com.procurement.notice.utils.toObject
import org.springframework.stereotype.Service
import java.time.LocalDateTime

@Service
interface EnquiryService {

    fun createEnquiry(cpid: String, stage: String, releaseDate: LocalDateTime, data: JsonNode): ResponseDto<*>

    fun addAnswer(cpid: String, stage: String, releaseDate: LocalDateTime, data: JsonNode): ResponseDto<*>

    fun unsuspendTender(cpid: String, stage: String, releaseDate: LocalDateTime, data: JsonNode): ResponseDto<*>
}

@Service
class EnquiryServiceImpl(private val releaseService: ReleaseService,
                         private val organizationService: OrganizationService,
                         private val releaseDao: ReleaseDao) : EnquiryService {

    companion object {
        private val SEPARATOR = "-"
        private val ENQUIRY_JSON = "enquiry"
        private val MS = "MS"
    }

    override fun createEnquiry(cpid: String, stage: String, releaseDate: LocalDateTime, data: JsonNode): ResponseDto<*> {
        val entity = releaseDao.getByCpIdAndStage(cpid, stage) ?: throw  ErrorException(ErrorType.DATA_NOT_FOUND)
        val enquiry = toObject(RecordEnquiry::class.java, toJson(data.get(ENQUIRY_JSON)))
        val record = toObject(Record::class.java, entity.jsonData)
        val ocId = record.ocid ?: throw ErrorException(ErrorType.OCID_ERROR)
        /*record*/
        record.id = getReleaseId(ocId)
        record.date = releaseDate
        record.tender.hasEnquiries = true
        /*add enquiry*/
        var enquiries = record.tender.enquiries ?: hashSetOf()
        if (enquiries.asSequence().none { it.id == enquiry.id }) {
            enquiries.add(enquiry)
            organizationService.processRecordPartiesFromEnquiry(record, enquiry)
            record.tender.enquiries = enquiries
            /*ms*/
            if (enquiries.size == 1) {
                val msEntity = releaseDao.getByCpIdAndStage(cpid, MS) ?: throw ErrorException(ErrorType.MS_NOT_FOUND)
                val ms = toObject(Ms::class.java, msEntity.jsonData).apply {
                    id = getReleaseId(cpid)
                    date = releaseDate
                    tag = listOf(Tag.COMPILED)
                    tender.hasEnquiries = true
                }
                releaseDao.saveRelease(releaseService.getMSEntity(cpid, ms))
            }
        }
        releaseDao.saveRelease(releaseService.getRecordEntity(cpid, stage, record))
        return getResponseDto(cpid, ocId)
    }

    override fun addAnswer(cpid: String, stage: String, releaseDate: LocalDateTime, data: JsonNode): ResponseDto<*> {
        val entity = releaseDao.getByCpIdAndStage(cpid, stage) ?: throw  ErrorException(ErrorType.DATA_NOT_FOUND)
        val enquiry = toObject(RecordEnquiry::class.java, toJson(data.get(ENQUIRY_JSON)))
        val record = toObject(Record::class.java, entity.jsonData)
        val ocId = record.ocid ?: throw ErrorException(ErrorType.OCID_ERROR)
        record.id = getReleaseId(ocId)
        record.date = releaseDate
        record.tender.enquiries?.asSequence()?.firstOrNull { it.id == enquiry.id }?.apply { this.answer = enquiry.answer }
                ?: throw ErrorException(ErrorType.ENQUIRY_NOT_FOUND)
        releaseDao.saveRelease(releaseService.getRecordEntity(cpid, stage, record))
        return getResponseDto(cpid, ocId)
    }

    override fun unsuspendTender(cpid: String, stage: String, releaseDate: LocalDateTime, data: JsonNode): ResponseDto<*> {
        val entity = releaseDao.getByCpIdAndStage(cpid, stage)
                ?: throw  ErrorException(ErrorType.DATA_NOT_FOUND)
        val record = toObject(Record::class.java, entity.jsonData)
        val ocId = record.ocid ?: throw ErrorException(ErrorType.OCID_ERROR)
        val dto = toObject(UnsuspendTenderDto::class.java, toJson(data))
        record.date = releaseDate
        record.id = getReleaseId(ocId)
        record.tender.statusDetails = dto.tender.statusDetails
        record.tender.tenderPeriod = dto.tender.tenderPeriod
        record.tender.enquiryPeriod = dto.tender.enquiryPeriod
        addAnswerToEnquiry(record.tender.enquiries, dto.enquiry)
        releaseDao.saveRelease(releaseService.getRecordEntity(cpid, stage, record))
        return getResponseDto(cpid, ocId)
    }

    private fun addAnswerToEnquiry(enquiries: HashSet<RecordEnquiry>?, enquiry: RecordEnquiry) {
        enquiries?.asSequence()?.firstOrNull { it.id == enquiry.id }?.apply {
            this.answer = enquiry.answer
        } ?: throw ErrorException(ErrorType.ENQUIRY_NOT_FOUND)
    }

    private fun getReleaseId(ocId: String): String {
        return ocId + SEPARATOR + milliNowUTC()
    }

    private fun getResponseDto(cpid: String, ocid: String): ResponseDto<*> {
        val jsonForResponse = createObjectNode()
        jsonForResponse.put("cpid", cpid)
        jsonForResponse.put("ocid", ocid)
        return ResponseDto(true, null, jsonForResponse)
    }
}